src/include/btree_inline.h

```diff
@@ -417,7 +417,7 @@ __wt_cache_page_byte_dirty_decr(WT_SESSION_IMPL *session, WT_PAGE *page, size_t
         /*
          * Take care to read the dirty-byte count only once in case we're racing with updates.
          */
-        WT_ORDERED_READ(orig, page->modify->bytes_dirty);
+        WT_READ_ONCE(orig, page->modify->bytes_dirty);
```

Comment:
I'm trying to figure out how to reason about whether these are correct.
I think we need to look through all code earlier and later to see whether any load into a variable that was done out-of-order with this assignment could cause an invalid execution path.

In most cases (including this one) there are atomic or other barrier protected operations nearby that are adding a level of certainty (or a bounds to how far the review needs to go).

My question: In figuring out whether the uses of `WT_READ_ONCE` here are valid, is it reasonable to include consideration about other concurrency control code in the vicinity? Put another way: do you think moving forward we (being all WiredTiger engineers) will reliably consider `WT_READ_ONCE` appearances when changing other stronger concurrency management mechanisms?

Reply:
It's a fair question.. I think there are two point of discussion.

> Put another way: do you think moving forward we (being all WiredTiger engineers) will reliably consider WT_READ_ONCE appearances when changing other stronger concurrency management mechanisms?

I don't see the need for special consideration of `READ_ONCE` in future. The new `READ_ONCE` will prevent splits and fuses at compile time, but beyond that `READ_ONCE` variable are no different from any other (shared) memory access. If removing adjacent barriers breaks something it's due to a mistake in our changed ordering logic, and if changing an adjacent atomic breaks something it's due to our usage of atomics regardless of the presence of `READ_ONCE`s.

> I think we need to look through all code earlier and later to see whether any load into a variable that was done out-of-order with this assignment could cause an invalid execution path.

The problem I see is that our current use of `ORDERED_READ` to achieve `READ_ONCE` requires a barrier, leaking ordering constraints into the surrounding code. 
Since all of the changes in this PR come with comments about only needing the `READ_ONCE` semantics we have a bit of leeway in assuming ordering is neither intentional or required, but I agree with you we need to be careful in case one of these `ORDERED_READ`s has become accidentally load bearing. 
When I reviewed I limited consideration to the loops surrounding each ORDERED_READ and was happy that there's no negative impact (this current example is a try-repeat loop, and later examples of the session walk are order independent), but I'm happy to re-review with this broader scope in mind.

This consideration will be needed each time we change an `ORDERED_READ` to a `READ_ONCE`, but once it's done I think we get the benefits described in the first paragraph and remove a bunch of complexity when dealing with `READ_ONCE` semantics in the future.